# HW1 – Sentiment Analysis on Twitter 
                                                                                                    @author:  Suchita Negi
                                                                                                    email id: snegi@scu.edu

## Assignment 1 : 
### To use keyword ‘Trump’ to collect the tweets and perform sentiment analysis.

In [1]:
# print all the outputs in a cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

### Data Collection:
The data with hashtag **Trump** has been collected using tweepy library in python to access twitter API.
The data is saved in text file Trump.txt which contains tweets with hashtag **#Trump**.

### Data Cleaning:

* The trump tweet data file is read using file i/o operators in python.
This data is then is then cleaned to remove unwanted words and pattern.
* **lower()** function is used to convert all tweets in each line to lower case.
* **replace()** function is also used to remove and replace unwanted notations/symbols; replaces the occurrences of a string with another.
* **split()** function splits the string in substrings; after the tweets is cleansed the data is split to be stored in a list.
* **re** is used to clean data with different patterns and substituted with simpler word for understanding. e.g : pattern like https:// is replaced by word URL.
Reference: https://docs.python.org/3/library/re.html

In [2]:
#Loading regular expression package to clean pattern in data.

import re


In [3]:
#Create an empty list to load the tweets with keyword Trump
tweet_trump=[]

#Open the text file in read mode 
f_tweet_trump = open('Trump.txt','r')

#Cleaning the data in each line of the file.
for line in f_tweet_trump:
    
#Convert all words to lower case and strip new line character .   
    line = line.lower()
    line = line.strip()
    
#CLeaning unwanted data pattern from each line.
    line = re.sub('((www\.[^\s]+)|(https?://[^\s]+))','URL',line)
    line = re.sub('@[^\s]+','UserName',line)
    line = re.sub('[\s]+', ' ', line)
    line = re.sub(r'#([^\s]+)', '', line)
    line = re.sub('((xe1[^\s]+)|(xe2[^\s]+)|(x80[^\s]+)|(xf0[^\s]+))','',line)
    
#Cleaning unwanted notations and symbols from each line of data file.   
    line = line.replace("rt","").replace(",","").replace(".","").\
    replace(":","").replace("!","").replace("$","").replace("?","").\
    replace("'","").replace('"',"").replace("\\n","").replace("n\'","").\
    replace("/","").replace("//","").replace("\\","").replace("|","").\
    replace("%","").replace("-","").replace("_","").replace("(","").replace(')',"")
    
#Split the words in each line and append each word to the list.
    for line in line.split():
        tweet_trump.append(line)
        
#Close the file in read mode
f_tweet_trump.close()

In [4]:
#Print to display 500 words from the tweets in the list.
print(tweet_trump[:500])

['b', 'UserName', 'the', 'cha', 'the', 'trump', 'campaign', 'should', 'be', 'plastering', 'absolutely', 'everywhereURL', 'URL', 'UserName', 'regardless', 'of', 'which', 'democrat', 'is', 'the', 'nominee', 'the', 'trump', 'campaign', 'is', 'ready', 'every', 'single', 'aspect', 'to', 'take', 'on', 'any', 'fight', 'we', 'UserName', 'see', 'list', 'of', 'president', 'trump', 'historical', 'and', 'miraculous', 'UserName', 'house', 'of', 'representatives', 'trump', 'has', 'been', 'impeached', 'trump', 'no', 'i', 'haven', 'this', 'is', 'literally', 'trump', 'impeachment', 'defense', 'UserName', 'UserName', 'UserName', 'UserName', 'UserName', 'UserName', 'UserName', 'UserName', 'URL', 'UserName', 'judicial', 'watch', 'files', 'foias', 'on', 'former', 'us', 'ambassador', 'to', 'ukraine', 'marie', 'yovanovitch', 'monitoring', 'of', 'president', 'trump', 'fa', 'on', 'which', 'us', 'senators', 'swear', 'impaiality', 'for', 'trump', 'impeachment', 'trial', 'bursts', 'into', 'flames', 'URL', 'UserNa

### Reference Word List:
To proceed with sentiment analysis we require a reference word list to compare our tweet words with.
Used reference text file in my directory for positive negative and stop words.
We read the below text file and assign these word to a list in python.
- positive.txt
- negative.txt
- stop.txt

In [5]:
#Reference positive word list.
#Convert all words in the file to lower case and append each reference word in the list.
positive_words =[]
f_positive= open('positive.txt','r')
for line in f_positive:
    line = line.lower()
    line = line.strip()
    positive_words.append(line)

f_positive.close()

In [6]:
#Reference negative word list
negative_words =[]
f_negative= open('negative.txt','r')
for line in f_negative:
    line = line.lower()
    line = line.strip()
    negative_words.append(line)

f_negative.close()

In [7]:
#Reference stop word list
stop_words =[]
f_stop= open('stop.txt','r')
for line in f_stop:
    line = line.lower()
    line = line.strip()
    stop_words.append(line)

f_stop.close()

In [8]:
#To analyze total number of reference words in each list
print("Total number of reference positive words  : %d " % len(positive_words))
print("Total number of reference negative words  : %d " % len(negative_words))
print("Total number of reference stop words  : %d " % len(stop_words))

Total number of reference positive words  : 2006 
Total number of reference negative words  : 4781 
Total number of reference stop words  : 319 


### Comparing the sentiments
We will compare each words in the tweet with reference words, whether positive negative or stop words are present in the tweet list.

- Initializing the counter for each sentiment
- Comparimg each word with respective word type (postive/negative/stop/others)
- Incrementing counter for each word type in tweets if matches with the respective word in reference word list.

**Our reference word list contains the word "trump" as a positive word.
As our search result is based on keyword "Trump" , our tweet will consist many trump words which can increase the positive sentiment count.
The true sentiment will not be reflected correctly , therefore we increase our positive word count excluding the word "trump" i.e. when word is not equal to "trump".**

In [9]:
#Counter Initialization
#positive word count
trump_positive_count=0
#negative word count
trump_negative_count=0
#stop word count
trump_stop_count= 0
#others word count
trump_others_count=0

#Comparing words in tweets to reference words and incrementing count accordingly

for word in tweet_trump:
    if word in positive_words and word!="trump":
        trump_positive_count += 1 
    elif word in negative_words:
        trump_negative_count += 1
    elif word in stop_words:
        trump_stop_count += 1
    else:
        trump_others_count += 1

### Sentiment Analysis:
We will Analyze the following:

**1. The total positive/negative/stop/other words used in the tweets.**

**2. Ratio of positive/negative/stop word/others compare to the whole word count.**

**3. Positive to Negative word Ratio.**


In [10]:
# Word counts in tweets

#Total number of words in the tweet list
trump_whole_word = len(tweet_trump)

print("Positive words count of tweets : %d" %trump_positive_count)

print("Negative words count of tweets : %d" %trump_negative_count)

print("Stop words count of tweets : %d" %trump_stop_count)

print("Other word count of tweets :  %d" %trump_others_count)

print("Whole (Total) word count of tweets : %d" %trump_whole_word)

print("positive v/s negative ratio = %.2f" %(trump_positive_count/trump_negative_count))



Positive words count of tweets : 871
Negative words count of tweets : 1196
Stop words count of tweets : 15883
Other word count of tweets :  21516
Whole (Total) word count of tweets : 39466
positive v/s negative ratio = 0.73


In [11]:
#Ratio of respecgtive words to total word count

print("Ratio of positive words to  whole word count = %.2f" %(trump_positive_count/trump_whole_word))

print("Ratio of negative words to  whole word count = %.2f" %(trump_negative_count/trump_whole_word))

print("Ratio of stop words to whole word count = %.2f" %(trump_stop_count/trump_whole_word))

print("Ratio of other words to whole word count = %.2f" %(trump_others_count/trump_whole_word))


Ratio of positive words to  whole word count = 0.02
Ratio of negative words to  whole word count = 0.03
Ratio of stop words to whole word count = 0.40
Ratio of other words to whole word count = 0.55


**4. The sum of all positive/negative sentiments observed.**

**5. The general sentiment whether negative or positive.**

**6. The sentiment whether weak or strong.**

In [12]:
#Total number of positive and negative sentiment in the tweets.
sum_pos_neg = trump_positive_count + trump_negative_count
print("Sum of all positive and negative words in tweets : %d" %(sum_pos_neg))

#difference between positive and negative words
diff_pos_neg = trump_negative_count - trump_positive_count
print("Difference in all positive and negative words in tweets : %d" %(diff_pos_neg))
                                                     
#Percentage of words that are positive w.r.t sum of positive negative words.
percent_positive = (trump_positive_count/sum_pos_neg) * 100

#Percentage of words that are negative w.r.t sum of positive negative words.
percent_negative = (trump_negative_count/sum_pos_neg) * 100

print("Percentage of words that are positive: %.2f%%" %(percent_positive))
print("Percentage of words that are negative: %.2f%%" %(percent_negative))

#Analyze general sentiment positive/negative and strong or weak.
if trump_positive_count>trump_negative_count:
    print("The general sentiment is Positive")
    if (percent_positive > 50):
        print("Positive sentiment is Strong as it is gretater than 50% as compared to negative sentiment.")
        print("Negative Sentiment is Weak")
else:
    print("The general sentiment is Negative")
    if (percent_negative > 50):
        print("Negative sentiment is Strong as it is gretater than 50% as compared to positive sentiment.")
        print("Positive Sentiment is Weak")



Sum of all positive and negative words in tweets : 2067
Difference in all positive and negative words in tweets : 325
Percentage of words that are positive: 42.14%
Percentage of words that are negative: 57.86%
The general sentiment is Negative
Negative sentiment is Strong as it is gretater than 50% as compared to positive sentiment.
Positive Sentiment is Weak


### Analysis:
In our analysis we observe that the negative word count used in the tweets(represents negative sentiment) are more than the positive sentiments.
The general sentiment in tweets for hashtag "trump" is negative and is around **57.86 %** compared to positive sentiments which is around 42.14 %.
Also the positive and negative count changes if we include the word trump as a positive word instead of name/hashtag in our tweet.
In conclusion sentiments towards this tweet is **strongly negative**.

## Assignment 2 : 
### To use keyword ‘Niners’ to collect the tweets and perform sentiment analysis.
We will analyze the twitter sentiments for San francisco 49ers (NINERS) football team who competes in NFL championships.

We will follow the same steps in our analysis for new keyword "Niners" as used above for keyword "Trump" .

- Data Collection
- Data Cleaning
- Reference word list
- Comapring the sentiments
- Sentiment Analysis

### Data Collection:
The data with hashtag **Niners** has been collected using tweepy library in python to access twitter API.
The data is saved in text file Niners.txt which contains tweets with hashtag **#Niners**.

### Data Cleaning

In [13]:
#Create an empty list to load the tweets with keyword Niners
tweet_niners=[]

#Open the text file in read mode 
f_tweet_niners = open('Niners.txt','r')

#Cleaning the data in each line of the file.
for line in f_tweet_niners:
    
#Convert all words to lower case and strip new line character from the file.   
    line = line.lower()
    line = line.strip()
#CLeaning unwanted data pattern from each line.
    line = re.sub('((www\.[^\s]+)|(https?://[^\s]+))','URL',line)
    line = re.sub('@[^\s]+','UserName',line)
    line = re.sub('[\s]+', ' ', line)
    line = re.sub(r'#([^\s]+)', '', line)
    line = re.sub('((xe1[^\s]+)|(xe2[^\s]+)|(x80[^\s]+)|(xf0[^\s]+))','',line)
    
#Cleaning unwanted notations and symbols from each line of data file.   
    line = line.replace("rt","").replace(",","").replace(".","").\
    replace(":","").replace("!","").replace("$","").replace("?","").\
    replace("'","").replace('"',"").replace("\\n","").replace("n\'","").\
    replace("/","").replace("//","").replace("\\","").replace("|","").\
    replace("%","").replace("-","").replace("_","").replace("(","").replace(')',"")
    
#Split the words in each line and append each word to the list.
    for line in line.split():
        tweet_niners.append(line)
        
#Close the file in read mode
f_tweet_niners.close()


In [14]:
#Print to display 500 words from the tweets in the list.
print(tweet_niners[:500])

['bdamn', 'bears', 'are', 'idiots', 'niners', 'have', 'raped', 'them', 'for', 'yearsbUserName', 'UserName', 'UserName', 'UserName', 'it', 'a', 'great', 'sb', 'matchup', 'two', 'passionate', 'fan', 'bases', 'th', 'URL', 'UserName', 'why', 'am', 'i', 'rooting', 'for', 'the', 'the', 'kyle', 'shanahanmike', 'shanahan', 'connection2', 'UserName', 'connection3', 'UserName', 'UserName', 'am', 'i', 'dreamingb', 'UserName', 'i', 'told', 'yall', 'id', 'get', 'yall', 'back', 'to', 'the', 'bowlrichard', 'sherman', 'stuck', 'to', 'his', 'word', 'the', 'niners', 'are', 'nfc', 'champs', 'via', 'UserName', 'UserName', 'barriers', 'going', 'to', 'be', 'broken', 'in', 'miamininers', 'offensive', 'assistant', 'katie', 'sowers', 'will', 'become', 'the', 'first', 'woman', 'to', 'coach', 'in', 'mf', 'niners', 'inna', 'superbowl', 'URL', 'good', 'thing', 'you', 'didn', 'pick', 'my', 'niners', 'but', 'won', 'be', 'long', 'till', 'the', 'drake', 'curse', 'happens', 'and', 'he', 'picks', 'us', 'i', 'be', 'hella

### Reference Word List

In [15]:
#Reference positive word list.
#Convert all words in the file to lower case and append each reference word in the list.
positive_words =[]
f_positive= open('positive.txt','r')
for line in f_positive:
    line = line.lower()
    line = line.strip()
    positive_words.append(line)

f_positive.close()

In [16]:
#Reference negative word list
negative_words =[]
f_negative= open('negative.txt','r')
for line in f_negative:
    line = line.lower()
    line = line.strip()
    negative_words.append(line)

f_negative.close()

In [17]:
#Reference stop word list
stop_words =[]
f_stop= open('stop.txt','r')
for line in f_stop:
    line = line.lower()
    line = line.strip()
    stop_words.append(line)

f_stop.close()

In [18]:
#To analyze total number of reference words in each list
print("Total number of reference positive words  : %d " % len(positive_words))
print("Total number of reference negative words  : %d " % len(negative_words))
print("Total number of reference stop words  : %d " % len(stop_words))

Total number of reference positive words  : 2006 
Total number of reference negative words  : 4781 
Total number of reference stop words  : 319 


### Comparing the sentiments
We will compare each words in the tweet with reference words, whether positive negative or stop words are present in the tweet list.


In [19]:
#Counter Initialization
#positive word count
niners_positive_count=0
#negative word count
niners_negative_count=0
#stop word count
niners_stop_count=0
#others word count
niners_others_count=0
for word in tweet_niners:
    if word in positive_words:
        niners_positive_count+=1 
    elif word in negative_words:
            niners_negative_count+=1
    elif word in stop_words:
            niners_stop_count+=1
    else:
        niners_others_count+=1

### Sentiment Analysis:
We will Analyze the following:

**1. The total positive/negative/stop/other words used in the tweets.**

**2. Ratio of positive/negative/stop word/others compare to the whole word count.**

**3. Positive to Negative word Ratio.**


In [20]:
# Word counts in tweets

#Total number of words in the tweet list
niners_whole_word = len(tweet_niners)

print("Positive words count of tweets : %d" %niners_positive_count)

print("Negative words count of tweets : %d" %niners_negative_count)

print("Stop words count of tweets : %d" %niners_stop_count)

print("Other word count of tweets :  %d" %niners_others_count)

print("Whole (Total) word count of tweets : %.2f" %niners_whole_word)

print("positive v/s negative ratio = %.2f" %(niners_positive_count/niners_negative_count))



Positive words count of tweets : 17
Negative words count of tweets : 23
Stop words count of tweets : 301
Other word count of tweets :  373
Whole (Total) word count of tweets : 714.00
positive v/s negative ratio = 0.74


In [21]:
#Ratio of respective words to total word count

print("Ratio of positive words to  whole word count = %.2f" %(niners_positive_count/niners_whole_word))

print("Ratio of negative words to  whole word count = %.2f" %(niners_negative_count/niners_whole_word))

print("Ratio of stop words to whole word count = %.2f" %(niners_stop_count/niners_whole_word))

print("Ratio of other words to whole word count = %.2f" %(niners_others_count/niners_whole_word))



Ratio of positive words to  whole word count = 0.02
Ratio of negative words to  whole word count = 0.03
Ratio of stop words to whole word count = 0.42
Ratio of other words to whole word count = 0.52


**4. The sum of all positive/negative sentiments observed.**

**5. The general sentiment whether negative or positive.**

**6. The sentiment whether weak or strong.**

In [22]:
#Total number of positive and negative sentiment in the tweets.
sum_pos_neg = niners_positive_count + niners_negative_count
print("Sum of all positive and negative words in tweets : %d" %(sum_pos_neg))

#difference between positive and negative words
diff_pos_neg = niners_negative_count - niners_positive_count
print("Difference in all positive and negative words in tweets : %d" %(diff_pos_neg))
                                                     
#Percentage of words that are positive w.r.t sum of positive negative words.
percent_positive = (niners_positive_count/sum_pos_neg) * 100

#Percentage of words that are negative w.r.t sum of positive negative words.
percent_negative = (niners_negative_count/sum_pos_neg) * 100

print("Percentage of words that are positive: %.2f%%" %(percent_positive))
print("Percentage of words that are negative: %.2f%%" %(percent_negative))

#Analyze general sentiment positive/negative and strong or weak.
if niners_positive_count>niners_negative_count:
    print("The general sentiment is Positive")
    if (percent_positive > 50):
        print("Positive sentiment is Strong as it is gretater than 50% as compared to negative sentiment.")
        print("Negative Sentiment is Weak")
else:
    print("The general sentiment is Negative")
    if (percent_negative > 50):
        print("Negative sentiment is Strong as it is gretater than 50% as compared to positive sentiment.")
        print("Positive Sentiment is Weak")



Sum of all positive and negative words in tweets : 40
Difference in all positive and negative words in tweets : 6
Percentage of words that are positive: 42.50%
Percentage of words that are negative: 57.50%
The general sentiment is Negative
Negative sentiment is Strong as it is gretater than 50% as compared to positive sentiment.
Positive Sentiment is Weak


### Analysis:
In our analysis we observe that the negative word count used in the tweets(represents negative sentiment) are more than the positive sentiments.
The general sentiment in tweets for hashtag "Niners" is negative and is around **57.50 %** compared to positive sentiments which is around 42.50 %.
In conclusion sentiments towards this tweet is **strongly negative**.